crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -201,55 +192,73 @@ pub async fn sync(
     }
 }
 
-/// Download and emit newly declared contract classes.
+/// Download and emit new contract classes.
+///
+/// New classes can come from:
+/// - DECLARE transactions
+/// - `old_declared_contracts` from the state diff (Cairo 0.x classes)
+/// - `declared_classes` from the state diff (Cairo 1.0 classes)
+/// - `deployed_contracts` from the state diff (DEPLOY transactions)
+/// - `replaced_classes` from the state diff
 ///
-/// We cannot remove the older way using `deploy_contracts` as this
-/// is required to handle older blocks which don't have declare transactions.
-async fn declare_classes(
+/// Note that due to an issue with the sequencer previously undeclared classes
+/// can show up in `replaced_classes`. This is caused by DECLARE v0 transactions
+/// that were _failing_ but the sequencer has still added the class to its list of
+/// known classes...
+async fn download_new_classes(
     block: &Block,
+    state_diff: &StateDiff,
     sequencer: &impl ClientApi,
     tx_event: &mpsc::Sender<Event>,
 ) -> Result<(), anyhow::Error> {
-    let declared_classes = block
-        .transactions
+    let deployed_classes = state_diff.deployed_contracts.iter().map(|x| x.class_hash);
+    let declared_cairo_classes = state_diff.old_declared_contracts.iter().cloned();
+    let declared_sierra_classes = state_diff
+        .declared_classes
         .iter()
-        .filter_map(|tx| {
-            use starknet_gateway_types::reply::transaction::DeclareTransaction;
-            use starknet_gateway_types::reply::transaction::Transaction::*;
-            match tx {
-                Declare(DeclareTransaction::V0(tx)) => Some(tx.class_hash),
-                Declare(DeclareTransaction::V1(tx)) => Some(tx.class_hash),
-                Declare(DeclareTransaction::V2(_)) => {
-                    todo!("v0.11.0: This maybe needs to be handled differently")
-                }
-                Deploy(_) | DeployAccount(_) | Invoke(_) | L1Handler(_) => None,
-            }
-        })
+        .map(|x| ClassHash(x.class_hash.0));
+    let classes_declared_in_transactions = block.transactions.iter().filter_map(|tx| {
+        use starknet_gateway_types::reply::transaction::DeclareTransaction;
+        use starknet_gateway_types::reply::transaction::Transaction::*;
+        match tx {
+            Declare(DeclareTransaction::V0(tx)) => Some(tx.class_hash),
+            Declare(DeclareTransaction::V1(tx)) => Some(tx.class_hash),
+            Declare(DeclareTransaction::V2(tx)) => Some(tx.class_hash),
```

Comment:
We would expect these to be present in the state diff, unless there is a bug in the gateway? So are we checking these just for safety?

In my mind, the `V0` is an exception to the rule, so maybe the rest shouldn't be included?

Reply:
I think _none_ of these is actually required, we're just keeping it around just to make sure? I'd fully expect all of these declared classes already be in the state update.

Reply:
Why do you think `v0` is an exception BTW?

Reply:
Ah I was confusing myself with v0 deployed contracts not already being declared for historical blocks.

However we should either delete the declare checks, or make a note of it. I imagine that would be a pretty big gateway bug (which we should discover if it exists).

Reply:
I'm OK with just removing them.